### PR TITLE
support multiple modifiers for binds

### PR DESCRIPTION
**Pull request by plsplsplslol**

### CODE REVIEW

**Comment by edg-l** [comment]
You are ignoring so many naming conventions...

File: src/game/client/components/binds.cpp [hunk 1]
```diff
@@ -66,10 +66,15 @@ void CBinds::Bind(int KeyID, const char *pStr, bool FreeOnly, int Modifier)
 		int Size = str_length(pStr) + 1;
 		m_aapKeyBindings[Modifier][KeyID] = (char *)malloc(Size);
 		str_copy(m_aapKeyBindings[Modifier][KeyID], pStr, Size);
-		if(Modifier)
-			str_format(aBuf, sizeof(aBuf), "bound %s+%s (%d) = %s", GetModifierName(Modifier), Input()->KeyName(KeyID), KeyID, m_aapKeyBindings[Modifier][KeyID]);
-		else
-			str_format(aBuf, sizeof(aBuf), "bound %s (%d) = %s", Input()->KeyName(KeyID), KeyID, m_aapKeyBindings[Modifier][KeyID]);
+		char aModifier[256] = { 0 };
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:69]:
```cpp
char aModifier[256];
aModifier[0] = '\0';
```
Fits more our style.

File: src/game/client/components/binds.cpp [hunk 2]
```diff
@@ -66,10 +66,15 @@ void CBinds::Bind(int KeyID, const char *pStr, bool FreeOnly, int Modifier)
...
+		for (int i = 1; i < MODIFIER_COUNT; i++) {
+			if (Modifier & (1 << i)) {
+				const char *ModifierName = GetModifierName(i);
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:72]: Pointers use the prefix `p`, therefore the name should be `pModifierName`

File: src/game/client/components/binds.cpp [hunk 3]
```diff
@@ -12,7 +12,7 @@ bool CBinds::CBindsSpecial::OnInput(IInput::CEvent Event)
 		int Mask = m_pBinds->GetModifierMask(Input());
 
 		bool ret = false;
-		for(int Mod = 0; Mod < MODIFIER_COUNT; Mod++)
+		for(int Mod = 0; Mod < (1 << MODIFIER_COUNT); Mod++)
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:15]: You use `(1 << MODIFIER_COUNT)` quite a lot, maybe create a constant and name it properly.

File: src/game/client/components/binds.cpp [hunk 4]
```diff
@@ -66,10 +66,15 @@ void CBinds::Bind(int KeyID, const char *pStr, bool FreeOnly, int Modifier)
 		int Size = str_length(pStr) + 1;
 		m_aapKeyBindings[Modifier][KeyID] = (char *)malloc(Size);
 		str_copy(m_aapKeyBindings[Modifier][KeyID], pStr, Size);
-		if(Modifier)
-			str_format(aBuf, sizeof(aBuf), "bound %s+%s (%d) = %s", GetModifierName(Modifier), Input()->KeyName(KeyID), KeyID, m_aapKeyBindings[Modifier][KeyID]);
-		else
-			str_format(aBuf, sizeof(aBuf), "bound %s (%d) = %s", Input()->KeyName(KeyID), KeyID, m_aapKeyBindings[Modifier][KeyID]);
+		char aModifier[256] = { 0 };
+		for (int i = 1; i < MODIFIER_COUNT; i++) {
+			if (Modifier & (1 << i)) {
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:71]: Brace style, change it to allman style

```
if()
{ // <- Notice the opening brace has its own line
}
```

File: src/game/client/components/binds.cpp [hunk 5]
```diff
@@ -66,10 +66,15 @@ void CBinds::Bind(int KeyID, const char *pStr, bool FreeOnly, int Modifier)
 		int Size = str_length(pStr) + 1;
 		m_aapKeyBindings[Modifier][KeyID] = (char *)malloc(Size);
 		str_copy(m_aapKeyBindings[Modifier][KeyID], pStr, Size);
-		if(Modifier)
-			str_format(aBuf, sizeof(aBuf), "bound %s+%s (%d) = %s", GetModifierName(Modifier), Input()->KeyName(KeyID), KeyID, m_aapKeyBindings[Modifier][KeyID]);
-		else
-			str_format(aBuf, sizeof(aBuf), "bound %s (%d) = %s", Input()->KeyName(KeyID), KeyID, m_aapKeyBindings[Modifier][KeyID]);
+		char aModifier[256] = { 0 };
+		for (int i = 1; i < MODIFIER_COUNT; i++) {
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:70]: Brace style

File: src/game/client/components/binds.cpp [hunk 6]
```diff
@@ -318,16 +316,22 @@ void CBinds::ConDumpBinds(IConsole::IResult *pResult, void *pUserData)
...
 					continue;
-				if(i)
-					str_format(aBuf, sizeof(aBuf), "%s+%s (%d) = %s", pBinds->GetModifierName(i), pBinds->Input()->KeyName(j), j, pBinds->m_aapKeyBindings[i][j]);
-				else
-					str_format(aBuf, sizeof(aBuf), "%s (%d) = %s", pBinds->Input()->KeyName(j), j, pBinds->m_aapKeyBindings[i][j]);
+
+				char aModifier[256] = { 0 };
+				for (int k = 1; k < MODIFIER_COUNT; k++) {
+					if (i & (1 << k)) {
+						const char *ModifierName = GetModifierName(k);
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:329]: Wrong variable name.

File: src/game/client/components/binds.cpp [hunk 7]
```diff
@@ -318,16 +316,22 @@ void CBinds::ConDumpBinds(IConsole::IResult *pResult, void *pUserData)
...
 				if(!pBinds->m_aapKeyBindings[i][j])
 					continue;
-				if(i)
-					str_format(aBuf, sizeof(aBuf), "%s+%s (%d) = %s", pBinds->GetModifierName(i), pBinds->Input()->KeyName(j), j, pBinds->m_aapKeyBindings[i][j]);
-				else
-					str_format(aBuf, sizeof(aBuf), "%s (%d) = %s", pBinds->Input()->KeyName(j), j, pBinds->m_aapKeyBindings[i][j]);
+
+				char aModifier[256] = { 0 };
+				for (int k = 1; k < MODIFIER_COUNT; k++) {
+					if (i & (1 << k)) {
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:328]: Brace style

File: src/game/client/components/binds.cpp [hunk 8]
```diff
@@ -318,16 +316,22 @@ void CBinds::ConDumpBinds(IConsole::IResult *pResult, void *pUserData)
...
 			{
 				if(!pBinds->m_aapKeyBindings[i][j])
 					continue;
-				if(i)
-					str_format(aBuf, sizeof(aBuf), "%s+%s (%d) = %s", pBinds->GetModifierName(i), pBinds->Input()->KeyName(j), j, pBinds->m_aapKeyBindings[i][j]);
-				else
-					str_format(aBuf, sizeof(aBuf), "%s (%d) = %s", pBinds->Input()->KeyName(j), j, pBinds->m_aapKeyBindings[i][j]);
+
+				char aModifier[256] = { 0 };
+				for (int k = 1; k < MODIFIER_COUNT; k++) {
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:327]: Brace style

File: src/game/client/components/binds.cpp [hunk 9]
```diff
@@ -421,17 +429,22 @@ void CBinds::ConfigSaveCallback(IConfig *pConfig, void *pUserData)
...
 		for(int j = 0; j < KEY_LAST; j++)
 		{
 			if(!pSelf->m_aapKeyBindings[i][j])
 				continue;
-			if(i)
-				str_format(aBuffer, sizeof(aBuffer), "bind %s+%s \"", pSelf->GetModifierName(i), pSelf->Input()->KeyName(j));
-			else
-				str_format(aBuffer, sizeof(aBuffer), "bind %s \"", pSelf->Input()->KeyName(j));
 
+			char aModifier[256] = { 0 };
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:439]: Style like I said previously.

File: src/game/client/components/binds.cpp [hunk 10]
```diff
@@ -421,17 +429,22 @@ void CBinds::ConfigSaveCallback(IConfig *pConfig, void *pUserData)
...
+			for (int k = 1; k < MODIFIER_COUNT; k++) {
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:440]: Brace style

File: src/game/client/components/binds.cpp [hunk 11]
```diff
@@ -421,17 +429,22 @@ void CBinds::ConfigSaveCallback(IConfig *pConfig, void *pUserData)
...
+				if (i & (1 << k)) {
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:441]: Brace style

File: src/game/client/components/binds.cpp [hunk 12]
```diff
@@ -421,17 +429,22 @@ void CBinds::ConfigSaveCallback(IConfig *pConfig, void *pUserData)
...
+					const char *ModifierName = GetModifierName(k);
```
Comment (edg-l) [inline src/game/client/components/binds.cpp:442]: Variable name

**Review by edg-l — CHANGES_REQUESTED** [review]
I don't think I like this or i find it useful, also you introduce repeated code.

**Comment by Learath2** [comment]
One problem with this is that it's now creating a rather large sparse array to store the binds in. I took a look at a couple other projects which seem to either use a hashmap or just put them all in a dense array. We could look into the performance implications of those.

**Comment by BannZay** [comment]
@Learath2 out of interest, it is possible to make hashmap work here? If we taking hash from key and its mod then I see no way to efficiently discover shift+b binding when user presses alt+shift+b.

P.s. I would prefer to not invoke shift+b on alt+shift+b press at the first point but this is current game behavior.

**Comment by Learath2** [comment]
> P.s. I would prefer to not invoke shift+b on alt+shift+b press at the first point but this is current game behavior.

It's not since the game doesn't even have multi modifier binds.

The check for the bind would be done when they hit the b, so it'd only ever trigger one.

**Comment by BannZay** [comment]
yes, user press alt+shift+b
if there is a binding 'shift+b' it will be triggered

**Comment by plsplsplslol** [comment]
@Learath2 should I continue fixing this or have someone else make this with a hashmap or dense array?

**Comment by Learath2** [comment]
This takes the memory usage for binds from 16k to 64k for a feature that will be seldom used. I think I like the feature but you should look into how you can get the memory usage down without a large performance hit.

**Comment by Learath2** [comment]
You can keep fixing this. I guess 64k is acceptable

**Comment by def-** [comment]
bors r+
Thanks!

**Pull request merged**